sdk/src/extras/disk.rs

```diff
@@ -0,0 +1,785 @@
+
+            let importer = super::DiskImport::try_from((self, progress_tx))?;
+            let handle = super::DiskImportHandle {
+                progress_rx,
+                cancel_tx,
+            };
+
+            Ok((importer.run_with_cancel(cancel_rx), handle))
+        }
+    }
+
+    impl<'a> TryFrom<(DiskImport<'a>, watch::Sender<u64>)> for super::DiskImport<'a> {
+        type Error = Error<types::Error>;
+
+        fn try_from(input: (DiskImport<'a>, watch::Sender<u64>)) -> Result<Self, Self::Error> {
+            let (builder, progress_tx) = input;
+
+            let project = builder.project.map_err(Error::InvalidRequest)?;
+            let description = builder.description.map_err(Error::InvalidRequest)?;
+            let upload_thread_ct = builder.upload_thread_ct.map_err(Error::InvalidRequest)?;
+            let disk = builder.disk.map_err(Error::InvalidRequest)?;
+            let disk_info = builder.disk_info.map_err(Error::InvalidRequest)?;
+            let image_info = builder.image_info.map_err(Error::InvalidRequest)?;
+
+            let upload_progress = Arc::new(AtomicU64::new(0));
+
+            Ok(Self {
+                client: builder.client,
+                project,
+                description,
+                upload_thread_ct,
+                disk,
+                disk_info,
+                image_info,
+                upload_progress,
+                progress_tx,
+                cleanup_started: AtomicBool::new(false),
+            })
+        }
+    }
+}
+
+// Upload to Nexus in 512k byte chunks
+const CHUNK_SIZE: u64 = 512 * 1024;
+
+/// Provides access to upload progress and cancellation for
+/// a disk import.
+#[derive(Debug)]
+pub struct DiskImportHandle {
+    progress_rx: watch::Receiver<u64>,
+    cancel_tx: oneshot::Sender<()>,
+}
+
+impl DiskImportHandle {
+    /// Returns watch channel `Receiver` that reports the current number of bytes uploaded.
+    /// For use with progress reports.
+    pub fn progress(&self) -> watch::Receiver<u64> {
+        self.progress_rx.clone()
+    }
+
+    /// Attempt to cancel the disk upload. If the upload has completed and the disk is
+    /// finalized then this will not remove the disk.
+    pub fn cancel(self) {
+        // Err indicates DiskImport has dropped and request has completed, nothing to do.
+        let _ = self.cancel_tx.send(());
+    }
+}
+
+#[derive(Debug)]
+struct DiskImport<'a> {
+    client: &'a Client,
+    project: NameOrId,
+    description: String,
+    upload_thread_ct: usize,
+    disk: Name,
+    disk_info: DiskInfo,
+    image_info: Option<ImageInfo>,
+    upload_progress: Arc<AtomicU64>,
+    progress_tx: watch::Sender<u64>,
+    cleanup_started: AtomicBool,
+}
+
+impl<'a> DiskImport<'a> {
+    async fn run_with_cancel(
+        self,
+        cancel_rx: oneshot::Receiver<()>,
+    ) -> Result<(), DiskImportError> {
+        self.check_for_existing_disk().await?;
+
+        let result = tokio::select! {
```

Comment:
do we want `biased;` here? i.e. I think we want to prioritize cancelation over progress.

Reply:
Good catch